File: backend/src/main/java/com/snackoverflow/toolgether/domain/user/controller/UserOauthController.java
package com.snackoverflow.toolgether.domain.user.controller;

import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.Multimap;
import com.snackoverflow.toolgether.domain.user.dto.request.AdditionalInfoRequest;
import com.snackoverflow.toolgether.domain.user.entity.User;
import com.snackoverflow.toolgether.domain.user.repository.UserRepository;
import com.snackoverflow.toolgether.domain.user.service.OauthService;
import com.snackoverflow.toolgether.global.dto.RsData;
import com.snackoverflow.toolgether.global.exception.custom.user.UserNotFoundException;
import com.snackoverflow.toolgether.global.filter.CustomUserDetails;
import com.snackoverflow.toolgether.global.filter.Login;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseCookie;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Mono;

import java.time.Duration;
import java.util.Map;

@Slf4j
@RestController
@RequiredArgsConstructor
public class UserOauthController {

    private final OauthService oauthService;
    private final UserRepository userRepository;

    @PostMapping("/login/oauth2/code/google")
    public RsData<?> handleAuthCode(@RequestBody Map<String, String> request,
                                    HttpServletResponse response) {
        try {
            String authCode = request.get("code");
            log.info("authCode = {}", authCode);
            if (authCode == null || authCode.isEmpty()) {
                throw new RuntimeException("authCode 값이 없습니다.");
            }

            // 토큰 가져오기
            Map<String, String> tokens = oauthService.getTokens(authCode);

            // 구글에서 발행한 리프레시 토큰을 쿠키에 저장
            setRefreshTokenCookie(response, tokens.get("refresh_token"));

            // 액세스 토큰에서 유저 정보 추출
            Map<String, Object> userInfo = oauthService.getUserInfo(tokens.get("access_token"), tokens.get("refresh_token"));
            String email = (String) userInfo.get("email");
            log.info("userInfo = {}", userInfo);

            Multimap<String, String> multiValueMap = ArrayListMultimap.create();
            multiValueMap.put("access_token", tokens.get("access_token"));

            /**
             * 소셜 로그인 사용자 존재 여부 체크
             * 존재하면 기존 로그인 진행 -> 액세스 토큰
             * 존재하지 않는 회원이라면 추가 정보 기입으로 이동
             */
            if (oauthService.existsUser(email)) {
                // 액세스 토큰을 응답 바디에 저장, PK 키 동시 전송
                User user = userRepository.findByEmail(email).orElseThrow(() -> new UserNotFoundException("사용자를 찾을 수 없습니다."));
                return new RsData<>(
                        "200-1",
                        "기존 사용자 로그인 성공",
                        Map.of("access_token", tokens.get("access_token"),
                                "user_id", user.getId(),
                                "nickname", user.getNickname())
                );
            }

            User socialUser = oauthService.createSocialUser(userInfo);
            return new RsData<>(
                    "201-1",
                    "신규 회원 가입 완료 - 추가 정보 입력 필요",
                    Map.of("additionalInfoRequired", socialUser.isAdditionalInfoRequired(),
                            "access_token", tokens.get("access_token"),
                            "user_id", String.valueOf(socialUser.getId()),
                            "nickname", socialUser.getNickname()
                    ));
        } catch (Exception e) {
            throw new RuntimeException("액세스 토큰 추출 중 오류 발생!");
        }
    }

    @PatchMapping("/oauth/users/additional-info")
    public Mono<RsData<Object>> updateAdditionalInfo(
            @Validated @RequestBody AdditionalInfoRequest request,
            @Login CustomUserDetails userDetails) {
        return oauthService.updateAdditionalInfo(userDetails.getEmail(), request)
                .map(isLocationValid -> {
                    if (Boolean.FALSE.equals(isLocationValid)) {
                        log.debug("isLocationValid: {}", isLocationValid);
                        return new RsData<>(
                                "400-1",
                                "위치 정보가 허용 범위를 벗어났습니다.",
                                null);
                    }

                    // 액세스 토큰 생성 후 바디로 전송
                    return new RsData<>(
                            "201-2",
                            "추가 정보가 등록되었습니다.",
                            null);
                });
    }

    @PostMapping("/oauth/token/refresh")
    public RsData<?> refreshAccessToken(@CookieValue("refresh_token") String refreshToken) {
        try {
            String newAccessToken = oauthService.refreshAccessToken(refreshToken);
            return new RsData(
                    "201-1",
                    "토큰 재발행 완료",
                    Map.of("access_token", newAccessToken)
            );
        } catch (Exception e) {
            return new RsData<>(
                    "401-1",
                    "리프레시 토큰이 유효하지 않습니다.",
                    null
            );
        }
    }

    private void setRefreshTokenCookie(HttpServletResponse response, String refreshToken) {
        ResponseCookie refreshCookie = ResponseCookie.from("refresh_token", refreshToken)
                .httpOnly(true)
                .secure(true) // HTTPS에서만 전송
                .path("/") // 특정 경로로 제한
                .maxAge(Duration.ofDays(30).toSeconds()) // 30일
                .sameSite("None")
                .build();

        response.addHeader("Set-Cookie", refreshCookie.toString());
        log.info("refreshToken - cookie 저장: " + refreshCookie.toString());
    }
}

File: backend/src/main/java/com/snackoverflow/toolgether/domain/user/controller/UserOauthController.kt
package com.snackoverflow.toolgether.domain.user.controller

import com.google.common.collect.ImmutableMap
import com.snackoverflow.toolgether.domain.user.dto.request.AdditionalInfoRequest
import com.snackoverflow.toolgether.domain.user.service.OauthService
import com.snackoverflow.toolgether.domain.user.service.UserServiceV2
import com.snackoverflow.toolgether.global.dto.RsData
import com.snackoverflow.toolgether.global.filter.CustomUserDetails
import com.snackoverflow.toolgether.global.filter.Login
import com.snackoverflow.toolgether.global.token.JwtService
import com.snackoverflow.toolgether.global.token.TokenService
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.Logger
import org.springframework.beans.factory.annotation.Value
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.RestController
import java.io.IOException

@RestController
class UserOauthController(
    @Value("\${custom.site.frontUrl}") private val frontUrl: String,
    @Value("\${jwt.access_expiration}") private val access: Long,
    @Value("\${jwt.refresh_expiration}") private val refresh: Long,
    private val oauthService: OauthService,
    private val tokenService: TokenService,
    private val userService: UserServiceV2,
    private val jwtService: JwtService,
    private val log: Logger
) {

    @GetMapping("/login/oauth2/code/google")
    fun handleSocialLogin(
        authCode: String,
        response: HttpServletResponse
    ): RsData<Any> {

        // 토큰 가져오기
        val accessToken = oauthService.getTokens(authCode)["access_token"] as String

        // 액세스 토큰으로 사용자 정보 가져오기
        val userInfo = oauthService.getUserInfo(accessToken) as Map<String, Any>
        val email = userInfo["email"] as String
        log.info("userInfo = {}", userInfo)

        // 사용자 존재 여부 체크
        return if (userService.existsUser(email)) {
            handleExistingUserLogin(email, response) // 기존에 존재하는 사용자는 로그인 처리
        } else {
            handleNewUserRegistration(userInfo, response) // 존재하지 않은 사용자는 추가 정보 입력
        }
    }

    // 휴대폰 검증 로직 (VerifyController) 이후에 -> 추가 정보 업데이트
    @PatchMapping("/oauth/users/additional-info")
    fun updateAdditionalInfo(
        @Validated request: AdditionalInfoRequest,
        @Login userDetails: CustomUserDetails
    ): RsData<Any> {

        oauthService.updateAdditionalInfo(userDetails.userEmail, request)

        return RsData("201-2", "추가 정보가 등록되었습니다.")
    }

    // JWT 토큰 추가 (access / refresh)
    private fun successLogin(
        email: String,
        userId: Long,
        response: HttpServletResponse
    ) {
        val (accessToken, refreshToken) = listOf(
            tokenService.createTokenByEmailAndId(email, userId, access),
            tokenService.createTokenByEmailAndId(email, userId, refresh)
        )

        response.apply {
            setHeader("Authorization", "Bearer $accessToken") // accessToken -> 헤더에 저장
            jwtService.setJwtSessionCookie(refreshToken, this) // refreshToken는 세션 쿠키로 저장 (페이지 닫을 시에 만료)
        }

        log.info("header 저장 토큰: {}, 쿠키 저장 토큰: {}", accessToken, refreshToken)
    }

    private fun handleExistingUserLogin(
        email: String,
        response: HttpServletResponse
    ): RsData<Any> {

        val user = userService.getUserByEmail(email)

        // 성공적인 로그인 처리
        successLogin(email, user.id!!, response)

        try {
            response.sendRedirect(frontUrl)
        } catch (e: IOException) {
            throw RuntimeException(e.message, e.cause)
        }

        return RsData(
            "200-1", "기존 사용자 로그인 성공",
            mapOf(
                "user_id" to user.id,
                "nickname" to user.nickname
            )
        )
    }

    private fun handleNewUserRegistration(
        userInfo: Map<String, Any>,
        response: HttpServletResponse
    ): RsData<Any> {

        val socialUser = oauthService.createSocialUser(userInfo)

        // 성공적인 로그인 처리
        successLogin(userInfo["email"] as String, socialUser.id!!, response)

        try {
            response.sendRedirect("$frontUrl/additional-info")
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

        return RsData(
            "201-1", "신규 회원 가입 완료 - 추가 정보 입력 필요",
            mapOf(
                "user_id" to socialUser.id,
                "nickname" to socialUser.nickname
            )
        )
    }
}

